src/server.lisp

```diff
@@ -158,90 +158,103 @@ DISPATCH-TABLE and LOGGING-STREAM are both required arguments.  TIMEOUT is of ty
+                             
+                             (setf reply (make-instance '|RPCReply|
+                                                        :|id| (|RPCRequest-id| request)
+                                                        :|result| result
+                                                        :|warnings| (coerce (reverse warnings) 'vector)))))
+                         (log-completion-message :info "Requested ~a completed" (|RPCRequest-method| request)))
+                     
+                     ;; this is where errors go where we can reply to the client
+                     (unknown-rpc-method (c)
+                       (declare (ignore c))
+                       (log-completion-message :err "Request ~a error: method ~a unknown"
+                                               (|RPCRequest-id| request)
+                                               (|RPCRequest-method| request))
+                       (setf reply (make-instance '|RPCError|
+                                                  :|id| (|RPCRequest-id| request)
+                                                  :|error| (format nil "Method named \"~a\" is unknown."
+                                                                   (|RPCRequest-method| request))
+                                                  :|warnings| (coerce (reverse warnings) 'vector))))
+                     (bt:timeout (c)
+                       (declare (ignore c))
+                       (log-completion-message :err "Request ~a error: timed out"
+                                               (|RPCRequest-id| request))
+                       (setf reply (make-instance '|RPCError|
+                                                  :|id| (|RPCRequest-id| request)
+                                                  :|error| (format nil "Execution timed out.  Note: time limit: ~a seconds." timeout)
+                                                  :|warnings| (coerce (reverse warnings) 'vector))))
+                     (error (c)
+                       (log-completion-message :err
+                                               "Request ~a error: Unhandled error in host program:~%~a"
+                                               (|RPCRequest-id| request)
+                                               c)
+                       (setf reply (make-instance '|RPCError|
+                                                  :|id| (|RPCRequest-id| request)
+                                                  :|error| (princ-to-string c)
+                                                  :|warnings| (coerce (reverse warnings) 'vector)))))
+                   
+                   ;; send the client response, whether success or failure
+                   (handler-case
+                       (%push-raw-request receiver identity empty-frame (serialize reply))
+                     (simple-error (c)
```

Comment:
should we do `error` here instead to grab all errors?

Reply:
Oh, yes, definitely. I actually didn't notice that I was using `error` right above this to do just that.